SSA/Experimental/Bits/Fast/BitStream.lean

```diff
@@ -373,12 +373,41 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
 
 local infix:20 " ≈ʷ " => EqualUpTo w
 
--- TODO: These sorries are difficult, and will be proven in a later Pull Request.
+-- TODO: This sorry is difficult, and will be proven in a later Pull Request.
 theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
   sorry
 
+theorem Nat.two_mod_mod (x : Nat) (w : Nat) (_ : 0 < w) : x % 2 ^ w % 2 = x % 2 := by
+  have y : 2 ^ 1 ∣ 2 ^ w := Nat.pow_dvd_pow 2 (by omega)
+  rw [pow_one] at y
+  exact Nat.mod_mod_of_dvd x y
+
+theorem Nat.parity_and (n m : Nat) :
+    2 ≤ n % 2 + m % 2 ↔ n % 2 = 1 ∧ m % 2 = 1 := by
+  omega
+
+theorem Nat.add_odd_iff_neq (n m : Nat) :
+    (n + m) % 2 = 1 ↔ (n % 2 = 1) ≠ (m % 2 = 1) := by
+  cases Nat.mod_two_eq_zero_or_one n
+  <;> cases Nat.mod_two_eq_zero_or_one m
+  <;> (
+    rename_i mparity nparity
+    simp [mparity, nparity, Nat.add_mod]
+  )
+
+theorem xor_decide (p q : Prop) [dp : Decidable p] [Decidable  q] :
+    (decide p).xor (decide q) = decide (p ≠ q) := by
+  cases dp
+  <;> (rename_i pt ; simp [pt])
+
 theorem ofBitVec_add : ofBitVec (x + y) ≈ʷ (ofBitVec x) + (ofBitVec y)  := by
-  sorry
+  intros n a
+  have add_lemma : ⟨BitVec.carry (n + 1) x y false , (x + y).getLsb n⟩ = (ofBitVec x).addAux (ofBitVec y) n := by
+    induction' n with n ih
+    · simp [addAux, BitVec.adcb, ofBitVec,a, BitVec.getLsb, Nat.two_mod_mod, BitVec.carry, ← Bool.decide_and, decide_eq_decide, xor_decide, Nat.parity_and x.toNat y.toNat, Nat.add_odd_iff_neq x.toNat y.toNat]
```

Comment:
Should maybe some of these be made to be simp lemmas if they are all required like this?

Reply:
Or, you could make simp lemmas for applied versions of some of the things you are unfolding (if there is an "obviously good" simplfication to be made)